### ISSUE TITLE
AppService Support

### ISSUE THREAD

**Issue by Mike-E-angelo** [dotnet/aspire#5675]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Aspire currently does not support AppService

### Describe the solution you'd like

AppService support. :)

### Additional context

This is in regard to this comment, which suggested to create an issue here to track:
https://devblogs.microsoft.com/dotnet/announcing-dotnet-aspire-8-2/?commentid=21117#comment-21117

Thank you for your effort and consideration. 🙏

**Comment by davidfowl** [comment]
We have no plans to support app service but I’ll leave this open for discussion.

**Comment by SophisticatedConsulting** [comment]
Would also appreciate app service support. 
Aspire is great. Container apps are great. 
However, for publishing a solution with multiple components (e.g. micro-services) to a test environment, container apps are very expensive ($$). App services should be supported as an alternative.

**Comment by mu-dawood** [comment]
Container apps are very expensive we really need  any alternative

**Comment by davidfowl** [comment]
First, this issue is about the **automagic** tooling that aspire provides that makes it easy to deploy to azure given the app model. If you continue to deploy manually, it will work, given that you provide the correct configuration in all of the different places. That said, let me just explain why we don't yet support app service and why container centric environments like ACA, AKS, docker compose are a better fit for what aspire provides. 

![Image](https://github.com/user-attachments/assets/8c4bfef1-36c7-4751-9f07-a0de7db37fdd)

The container app environment serves as a default security boundary and services that run inside of the environment. Containers can run inside of the environment and can expose multiple protocols (tcp, http etc). Service discovery is setup via envoy for both http and https, http2 is supported (so grpc just works). Services need to opt-in to being exposed to the public internet. 

Communication between ACA services and azure services is over the public internet right now, but we plan to take advantage of ACA's implementation to setup secure communication between services in the cluster and any declared services in azure.

Components baked into ACA are in blue (the aspire dashboard, the resource server, and the otel collector, data protection service). These systems are required to make the aspire experience work well (the reason why distributed traces just work when you deploy).

This not to say that that App Service cannot work in some very constrained scenarios, or, if you only use certain resources but it's less than idea. On top of that, the dashboard experience would need to be designed, the security model that app service has today doesn't lend itself towards the "internal by default" mantra we have without additional infrastructure. We're discussion what an "ideal" environment looks like with the app service team to see if it makes sense to invest in the work required to make it work in a similar way.

**Comment by Mike-E-angelo** [comment]
Thank you for that context @davidfowl, it is appreciated.  For me, there is the consideration of an Azure feature that has been around for nearly a decade.  There's a great deal of familiarity with this and therefore value.  However, others have mentioned pricing which is just as important.  Indeed, after applying Aspire as designed with a basic SQL server + container setup, the billing was over $1,000 for a month of usage on my Azure account.

This seems beyond extreme and completely unviable, especially for hobbyists where you will see a good deal of potential adoption.  In my case, I had Microsoft for Startups credits to cover the costs, so no harm to my pocket.  Still, I unfortunately cannot use (or recommend) Aspire at the moment due to the lack of support for AppService and the resulting costs with a default setup.

Additionally, please note that I had a client who was interested in using Aspire but lost interest after reporting its associated costs to them.  Even those with money find $1k+/mo for a basic web hosting setup a little rich for their tastes.  Thank you for any consideration toward addressing this concern.

**Comment by JennyLawrance** [comment]
@Mike-E-angelo We can investigate the $1000 bill you are referring to. Can you send the Azure resource details in question? (Even if it was deleted that should be ok as long it was in the past 60 days). Please send us the resource details to acasupport(at)microsoft(dot)com

**Comment by elietebchrani** [comment]
I strongly agree with this feature request. While Container Apps provides a great deployment option for Aspire applications, having Azure App Service as an alternative would be extremely valuable for several reasons:

1. Many developers already have experience deploying to App Service and understand its management model.
2. App Service offers a more straightforward deployment experience without container-specific concerns.
3. For smaller projects and development environments, App Service can be significantly more economical than Container Apps.
4. Reduced infrastructure management overhead compared to container-based solutions.

For teams working on smaller to medium-sized applications or those with budget constraints, App Service support would make Aspire more accessible and practical. This addition would greatly enhance Aspire's flexibility while maintaining its core benefits of simplified cloud-native deployment.

**Comment by davidfowl** [comment]
We're looking into app service support with the app service team, still early and investigating so I don't have any timelines to promise/discuss, this work is exploratory.

> App Service offers a more straightforward deployment experience without container-specific concerns.

Any experience aspire builds with app service will be based on app service using container images. This is how azd works with aspire today and what we would continue to do to avoid any ties to app service specific apis during building and pushing applicaitons.

**Comment by davidfowl** [comment]
This is the first cut of app service support, no timeline for when this ships yet, but we're working with the app service team to make a subset of scenarios work https://github.com/dotnet/aspire/pull/9090.

**Comment by Mike-E-angelo** [comment]
That's awesome, @davidfowl!  Thank you so much for your efforts in this area.  I enjoyed Aspire when I used it, and would appreciate having it in all my (AppService) solutions, and I know others would, too. 😊🙏

**Comment by davidfowl** [comment]
We shipped a preview of this in 9.3 closing this issue as there are other work items for moving this from preview to GA over the next several releases of aspire.

https://azure.github.io/AppService/2025/05/19/Aspire-on-App-Service.html

**Comment by Mike-E-angelo** [comment]
So awesome @davidfowl and team!  Thank you so much for listening to your customers and providing such great value with your product.  I have this on my radar to integrate once the v2 of my product is in place, hopefully within the next few months. 😊🙏✨
